# Lab 8: Regular Expressions
In this lab assignment you will be practicing constructing regular expressions. First you will try your hand at some basic regular expressions in order to match words and numbers in a text file. Then you’ll try some more complex regular expressions in "real" situations.

For all of the exercises, you will use only the basic regular expressions (BRE) discussed in class. The only symbols you can use are:

    *     Zero or more
    +     One or more
    ?     Zero or one / Optional
    [X-Y] One character out the range X to Y
    .     Any one character
    |     Or
    ( )   Grouping
    { }   Repetition
    \.    Dot
    \$    Dollar sign
    \     Escape any other special character

Some common regular expression patterns are:

    [0-9]     Any one digit
    [0-9]+    One or more digits
    [a-zA-Z]  Any one letter (upper or lower case)
    [a-z]{3}  Three lower case letters
    -?[0-9]+  A number with an optional negative sign
    .*        Any number of any characters (including no characters at all)
    
## How to Do This Assignment

Before starting, upload the four text files for this assignment:

* contacts.txt
* program.txt
* perl.txt
* weblog.txt

In each problem function you'll see the string TODO. Replace TODO with the regular expression needed to do each problem.

There are no automated checks for your solutions; please eyeball them yourself to visually verify the output.

In [1]:
# Run this cell to define the function you'll need for the rest of the assignment.
#WRITTEN FOR US_EXECUTE ONLY
from re import finditer

# Return an array of strings that match the given
# regular expression
def extract(regexp, fname):
    matches = []
    with open(fname) as f:
        for line in f:
            for match in finditer(regexp, line):
                matches.append(match.group())
    f.close()
    return matches

# Words and Numbers
Take a look at the file named **contacts.txt**. It’s a list of fake names and addresses.

## Problem 1
Extract all numbers in the text. Numbers are simply one or more consecutive digits. It's OK if the numbers that get extracted are in the middle of email addresses.

**Check:** The first few numbers should be **1116, 42, 02919, 401, 783, 9567, 5361, 40108, ...**

In [4]:
extract(r'[0-9]+', "contacts.txt")

['1116',
 '42',
 '02919',
 '401',
 '783',
 '9567',
 '5361',
 '40108',
 '3837',
 '502',
 '269',
 '4969',
 '9088',
 '5',
 '33912',
 '1923',
 '305',
 '636',
 '5462',
 '7320',
 '03907',
 '207',
 '457',
 '6646',
 '4293',
 '425',
 '83342',
 '208',
 '860',
 '6627',
 '5245',
 '736',
 '95652',
 '0736',
 '719',
 '332',
 '4698',
 '5331',
 '26412',
 '3383',
 '304',
 '388',
 '6503',
 '2',
 '839',
 '357',
 '19301',
 '34',
 '741',
 '96014',
 '720',
 '523',
 '0191',
 '681',
 '30188',
 '404',
 '564',
 '5544',
 '303',
 '20012',
 '1118',
 '202',
 '467',
 '3939',
 '695',
 '852',
 '59469',
 '0027',
 '50',
 '915',
 '72036',
 '501',
 '745',
 '9171',
 '47',
 '686',
 '95117',
 '4431',
 '719',
 '678',
 '1479',
 '2675',
 '019',
 '58231',
 '3992',
 '701',
 '146',
 '2174',
 '2168',
 '124',
 '30216',
 '228',
 '769',
 '7043']

## Problem 2
Extract all words in the text. Words are composed of one or more letters (upper and lower case). Words can be parts of email addresses, too.

**Check:** The first few words are **Chong, Merrih, Unit, G, Cranston, RI, merrihchong, gmail, com, ...**

In [7]:
extract(r'[a-zA-Z][a-zA-Z]+', "contacts.txt")

['Chong',
 'Merrih',
 'Unit',
 'Cranston',
 'RI',
 'merrihchong',
 'gmail',
 'com',
 'Verona',
 'McRome',
 'Leanna',
 'Crescent',
 'Brandenburg',
 'KY',
 'verona',
 'brandenburg',
 'ci',
 'ky',
 'us',
 'Eunice',
 'Hausladen',
 'Hangar',
 'Fort',
 'Myers',
 'FL',
 'eunice',
 'yahoo',
 'com',
 'Cristie',
 'Coblentz',
 'Terrell',
 'Shores',
 'Ogunquit',
 'ME',
 'ccobl',
 'bbc',
 'co',
 'uk',
 'Brendon',
 'Wiederwax',
 'Jerald',
 'Street',
 'Apt',
 'Naf',
 'ID',
 'brendon',
 'snarky',
 'me',
 'Amie',
 'Taibl',
 'Khalil',
 'Drive',
 'McClellan',
 'Afb',
 'CA',
 'taibl',
 'hotmail',
 'com',
 'Rosette',
 'Bierkortte',
 'Margarett',
 'Groves',
 'Orlando',
 'WV',
 'rbier',
 'psolutions',
 'com',
 'William',
 'Veeser',
 'Remington',
 'Well',
 'Suite',
 'Paoli',
 'PA',
 'veeser',
 'yahoo',
 'com',
 'Corrine',
 'Officer',
 'Bergnaum',
 'Valley',
 'Callahan',
 'CA',
 'officer',
 'gmail',
 'com',
 'Dan',
 'Layer',
 'Alek',
 'Glen',
 'Woodstock',
 'GA',
 'dan',
 'layer',
 'and',
 'cabinets',
 'com',


## Problem 3
Extract all five- and nine-digit ZIP codes. Keep in mind that nine-digit zip codes have a hyphen in them.

**Check:** The first few ZIP codes are **02919, 40108-3837, 33912-1923, 03907, ...**

In [8]:
extract(r'\b\d{5}(?:-\d{4})?\b', "contacts.txt")

['02919',
 '40108-3837',
 '33912-1923',
 '03907',
 '83342',
 '95652-0736',
 '26412-3383',
 '19301',
 '96014',
 '30188',
 '20012-1118',
 '59469-0027',
 '72036',
 '95117-4431',
 '58231-3992',
 '30216']

## Problem 4
Extract all email addresses. For our purposes, an email address is a *username*, followed by an @ symbol, followed by the *hostname*.  
The username consists of one or more letters, numbers, dots, dashes, and underscores.  
The hostname consists of one or more letters, numbers, dots, and dashes.

**Check:** Here are the first few email address: `merrihchong@gmail.com`, `verona@brandenburg.ci.ky.us`, `eunice_h@yahoo.com`, `ccobl@bbc.co.uk`, `brendon@snarky.me`, `taibl@hotmail.com`

In [9]:
extract(r'[a-zA-Z0-9\.\-_]+@[a-zA-Z0-9\.\-]+', "contacts.txt")

['merrihchong@gmail.com',
 'verona@brandenburg.ci.ky.us',
 'eunice_h@yahoo.com',
 'ccobl@bbc.co.uk',
 'brendon@snarky.me',
 'taibl@hotmail.com',
 'rbier@p2psolutions.com',
 'veeser34@yahoo.com',
 'officer.c@gmail.com',
 'dan.layer@r-and-r-cabinets.com',
 'knitting.lady@gmail.com',
 'wes_wes@corp.yahoo.com']

# Tokenizing

Take a look at **program.txt**. This is a simple computer program from a non-existent programming language. In this language, keywords are uppercase and everything else is lowercase.

*Tokenizing* is the phase of compilation that identifies the parts of the program in preparation for the parser. The lexer needs to scan through the source code, locate the tokens, and tag them as such. Here, you will write regular expressions (one per problem) to find various things to be tokenized.

## Problem 5
Extract all the *numbers* in the program. Keep in mind they could be a decimal amount and may be negative.

**Check:** Here are the numbers you should find: 1.8, 32, -2.547, 7, 20.0

In [14]:
extract(r'-?\b-?[0-9]+(?:\.[0-9]+)?\b', "program.txt")

['1.8', '32', '-2.547', '7', '20.0']

## Problem 6
Extract all the *identifiers*. Identifiers are the names that programmers give to things, such as variables, function names, etc. In this programming language, they are composed of lowercase letters.

The identifiers in the program are: **project, x, y, age, convert, temp, zip, foo,** ...

In [15]:
extract(r'\b[a-z]+\b', "program.txt")

['project',
 'x',
 'y',
 'age',
 'convert',
 'temp',
 'zip',
 'temp',
 'foo',
 'zip',
 'kipper',
 'foo',
 'kipper',
 'age',
 'x',
 'age',
 'x',
 'y',
 'x',
 'convert',
 'x',
 'y',
 'convert',
 'x',
 'y']

# Tokenizing a Real Language

The next file is **perl.txt**. This is a real program written in the Perl programming language. It reads in a comma-separated file (CSV) in which each line is a student’s last name, first name, and ID number. It converts each line into an array containing the username, password, first name, and last name — enough information to make user accounts on a server.

## Problem 7
Extract the *variables*. In Perl, variables begin with a $ or @ symbol. The next character can be a letter or underscore. The rest of the variable name can be letters, digits, or underscores. In Perl, @_ is a valid variable, too!

In this program, some of the variables are: `$lname`, `$fname`, `$sid`, `$LIST_SEP`, `@line`, `$first4`, `$full_name`, `@_`, `$password`, ...

In [25]:
extract(r'\$[a-zA-Z_][a-zA-Z0-9_]*|\@[a-zA-Z_][a-zA-Z0-9_]*', "perl.txt")

['$lname',
 '$fname',
 '$sid',
 '$LIST_SEP',
 '@line',
 '@line',
 '$fname',
 '$lname',
 '@line',
 '$fname',
 '$lname',
 '$sid',
 '@line',
 '$fname',
 '@line',
 '$lname',
 '@line',
 '$fname',
 '$lname',
 '@_',
 '$fname',
 '$lname',
 '$fname',
 '$lname',
 '$sid',
 '@_',
 '$full_name',
 '$lname',
 '$fname',
 '$first4',
 '$full_name',
 '$password',
 '$first4',
 '$sid',
 '$password']

## Problem 8
Extract the *comments*. Comments begin with a # symbol and continue to the end of the line. They can contain any number of any characters, including none.

Some examples from the program:

    #! /usr/bin/perl
    # Convert CSV file with student info into a file
    # that can be used to make accounts.
    #
    # Output is username, password, first name, last name
    # Change array element separator to comma
    # Print the whole array
    # Return first initial, last name, all lower case

In [37]:
extract(r'#.*?(?=\n|$)', "perl.txt")

['#! /usr/bin/perl',
 '# Convert CSV file with student info into a file',
 '# that can be used to make accounts.',
 '#',
 '# Output is username, password, first name, last name',
 '# Change array element separator to comma',
 '# Print the whole array',
 '# Return first initial, last name, all lower case',
 '# Make password out of first name, last name, and student ID.',
 '# Concatenate last name and first name. Convert to lower case.',
 '# Truncate to 4 chars.',
 '# Append last 4 digits of student ID']

# Web Server Log Files

Web servers log every request that comes in. A request is an individual document, such as an HTML file, an image, a JavaScript program, CSS file, and so forth. The last file, **weblog.txt**, contains a portion of a typical log file containing, among other data points:

* the IP address of the web browser
* the date/time of the request
* the file requested, status code
* the file size

## Problem 9
Extract the IP addresses. IP addresses are four numbers separated by dots. Remember that a dot is a special character, so you may need to *escape* it by using a backslash: `\.`

In [51]:
extract(r'\b(?:\d{1,3}\.){3}\d{1,3}\b', "weblog.txt")

['128.242.249.12',
 '64.78.116.124',
 '105.12.94.91',
 '180.76.5.156',
 '174.50.148.8',
 '208.115.111.75',
 '199.21.99.72',
 '208.115.113.91',
 '71.142.241.232',
 '76.246.40.237',
 '76.246.63.198',
 '199.21.99.72',
 '76.246.63.198']

## Problem 10
Extract the dates and times, which look like this: **06/May/2019:03:45:36**

In [64]:
extract(r'\b\d{2}/[a-zA-Z]+/\d{4}:\d{2}:\d{2}:\d{2}\b', "weblog.txt")

['30/Jan/2019:19:23:31',
 '02/Feb/2019:03:45:37',
 '02/Feb/2019:04:32:22',
 '16/Feb/2019:15:29:19',
 '21/Feb/2019:20:05:44',
 '06/Mar/2019:17:14:32',
 '08/Mar/2019:08:16:21',
 '29/Mar/2019:11:08:42',
 '10/Apr/2019:22:48:09',
 '15/Apr/2019:08:32:30',
 '04/May/2019:10:37:30',
 '06/May/2019:02:16:55',
 '06/May/2019:11:25:32']

## Problem 11
Extract the image filenames. Images end with **.jpg** or **.png**. Filename can contains letters, numbers, and underscores.

Just extract the filename, not the whole path. For example, we want **header.jpg**, but not **/sites/all/themes/litejazz/js/header.jpg**.

You should get: header.jpg, greetings_art.jpg, backgnd.jpg, footer2.png

In [78]:
extract(r'(?<=/)[\w\d_]+\.(?:jpg|png)\b', "weblog.txt")

['header.jpg', 'greetings_art.jpg', 'backgnd.jpg', 'footer2.png']